src/Microsoft.AspNetCore.Sockets/Transports/ServerSentEventsTransport.cs

```diff
@@ -33,33 +35,28 @@ public ServerSentEventsTransport(ReadableChannel<Message> application, ILoggerFa
             await context.Response.Body.FlushAsync();
 
             var pipe = context.Response.Body.AsPipelineWriter();
+            var output = new PipelineTextOutput(pipe, TextEncoder.Utf8); // We don't need the Encoder, but it's harmless to set.
 
             try
             {
                 while (await _application.WaitToReadAsync(token))
                 {
-                    var buffer = pipe.Alloc();
                     while (_application.TryRead(out var message))
                     {
-                        if (!ServerSentEventsMessageFormatter.TryFormatMessage(message, buffer.Memory.Span, out var written))
+                        if (!ServerSentEventsMessageFormatter.TryWriteMessage(message, output))
                         {
-                            // We need to expand the buffer
-                            // REVIEW: I'm not sure I fully understand the "right" pattern here...
-                            buffer.Ensure(LongPollingTransport.MaxBufferSize);
+                            // We ran out of space to write, even after trying to enlarge.
+                            // This should only happen in a significant lack-of-memory scenario.
 
-                            // Try one more time
-                            if (!ServerSentEventsMessageFormatter.TryFormatMessage(message, buffer.Memory.Span, out written))
-                            {
-                                // Message too large
-                                throw new InvalidOperationException($"Message is too large to write. Maximum allowed message size is: {LongPollingTransport.MaxBufferSize}");
-                            }
+                            // IOutput doesn't really have a way to write incremental
+
+                            // REVIEW: I really am not sure what exception to throw here... OOM seems right, but it feels so wrong to throw it ourselves...
+                            throw new OutOfMemoryException("Ran out of space to format messages!");
                         }
-                        buffer.Advance(written);
-                        buffer.Commit();
-                        buffer = pipe.Alloc();
-                    }
 
-                    await buffer.FlushAsync();
+                        // REVIEW: Flushing after each message? Good? Bad? We can't access Commit because it's hidden inside PipelineTextOutput
```

Comment:
Haven't heard anything from @davidfowl . Flushing after every message is reasonable, it just might be overzealous. I'll leave it as such. Waiting until the whole batch is done before flushing means we can't release any buffers until the entire batch is done, which seems non-ideal.